### PR TITLE
dedup prebuild triggers

### PR DESCRIPTION
**Pull request by AlexTugarev**
## Related Issue(s)

Fixes #

## Release Notes

```release-note
```

## Werft options:

- [x] /werft with-preview
- [x] /werft with-clean-slate-deployment
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`

### CODE REVIEW

**Comment by roboquat** [comment]
@AlexTugarev: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by AlexTugarev** [comment]
Screen Shot 2022-09-22 at 13 59 20
Screen Shot 2022-09-22 at 13 59 34

can reproduce it, finally! that was a long way.

File: components/server/install-gh-app.sh [hunk 1]
```diff
@@ -21,7 +21,7 @@ fi
 
 # turn spaces into newlines, in case the key got pasted in on the env var dashboard interface
 TMP=${GH_APP_KEY// RSA PRIVATE /RSA_PRIVATE}
-TMP2=${TMP// /$'\n'}
+TMP2=${TMP//\\n/$'\n'}
```
Comment (AlexTugarev) [inline components/server/install-gh-app.sh:24]: todo: revert

File: components/server/ee/src/user/eligibility-service.ts [hunk 2]
```diff
@@ -72,41 +72,41 @@ export class EligibilityService {
...
-                // That violates the API contract: `student` and `faculty` need to be mutually exclusive
-                log.warn(
-                    logCtx,
-                    "result of GitHub Eduction API violates the API contract: student and faculty need to be mutually exclusive",
-                    result,
-                );
-                return { student: false, faculty: false };
-            }
-            return result;
+            // const rawResponse = await fetch("https://education.github.com/api/user", {
```
Comment (AlexTugarev) [inline components/server/ee/src/user/eligibility-service.ts:85]: todo: revert

File: components/gitpod-protocol/src/encryption/encryption-engine.spec.ts [hunk 3]
```diff
@@ -24,7 +24,7 @@ class TestEncryptionEngineImpl {
 
     @test basicSymmetry() {
         const plaintext = "12345678901234567890";
-        const key = new Buffer(this.testkey, "base64");
+        const key = Buffer.from(this.testkey, "base64");
```
Comment (AlexTugarev) [inline components/gitpod-protocol/src/encryption/encryption-engine.spec.ts:27]: todo: extract to small fix PR. this is about annoying errors logged.

File: components/server/ee/src/prebuilds/github-app.ts [hunk 4]
```diff
@@ -487,6 +530,7 @@ export class GithubApp {
 
     protected async onPrStartPrebuild(
         tracecContext: TraceContext,
+        event: WebhookEvent,
```
Comment (AlexTugarev) [inline components/server/ee/src/prebuilds/github-app.ts:533]: todo: revert line

File: components/server/ee/src/prebuilds/github-app.ts [hunk 5]
```diff
@@ -423,14 +435,45 @@ export class GithubApp {
             user = r.user;
             project = r.project;
 
-            const prebuildStartPromise = await this.onPrStartPrebuild({ span }, ctx, config, context, user, project);
-            if (prebuildStartPromise) {
-                await this.onPrAddCheck({ span }, config, ctx, prebuildStartPromise);
+            await this.webhookEvents.updateEvent(event.id, {
```
Comment (AlexTugarev) [inline components/server/ee/src/prebuilds/github-app.ts:438]: todo: extract to PR about missing 'pulls' events

File: components/server/ee/src/prebuilds/github-app.ts [hunk 6]
```diff
@@ -181,12 +182,17 @@ export class GithubApp {
             options.getRouter("/reconfigure").get(
                 "/",
                 asyncHandler(async (req: express.Request, res: express.Response) => {
-                    const gh = await app.auth();
-                    const data = await gh.apps.getAuthenticated();
-                    const slug = data.data.slug;
-
-                    const state = req.query.state;
-                    res.redirect(`https://github.com/apps/${slug}/installations/new?state=${state}`);
+                    try {
```
Comment (AlexTugarev) [inline components/server/ee/src/prebuilds/github-app.ts:185]: todo: extract in simple PR to log the actual error *and* stop rending the stack trace

File: components/server/src/workspace/workspace-starter.ts [hunk 7]
```diff
@@ -720,28 +719,39 @@ export class WorkspaceStarter {
...
                     prebuild.error = err.toString();
 
                     await this.workspaceDb.trace({ span }).storePrebuiltWorkspace(prebuild);
-                    await this.messageBus.notifyHeadlessUpdate({ span }, workspace.ownerId, workspace.id, <
-                        HeadlessWorkspaceEvent
-                    >{
+                    await this.messageBus.notifyHeadlessUpdate({ span }, workspace.ownerId, workspace.id, {
                         type: HeadlessWorkspaceEventType.Failed,
-                        // TODO: `workspaceID: workspace.id` not needed here? (found in ee/src/prebuilds/prebuild-queue-maintainer.ts and ee/src/bridge.ts)
+                        workspaceID: workspace.id, // required at least in PrebuildStatusMaintainer
```
Comment (AlexTugarev) [inline components/server/src/workspace/workspace-starter.ts:747]: todo: search logs for "cannot read workspaceID of undefined", this looks like the reason for failing PrebuildStatusMaintainer calls. worth a small PR

**Comment by AlexTugarev** [comment]
see in logs `A prebuild is already running for this commit. (2)`

🎉  yeah!

Screen Shot 2022-09-22 at 15 34 30

File: components/server/ee/src/workspace/workspace-factory.ts [hunk 8]
```diff
@@ -168,6 +174,27 @@ export class WorkspaceFactoryEE extends WorkspaceFactory {
...
+                        commitContext.repository.cloneUrl,
+                        CommitContext.computeHash(commitContext),
+                    );
+                log.info("createForStartPrebuild–2", {
+                    foundPrebuild: !!existingPWS,
+                    context,
+                    cloneUrl: commitContext.repository.cloneUrl,
+                    commit: CommitContext.computeHash(commitContext),
+                });
+                if (existingPWS) {
```
Comment (AlexTugarev) [inline components/server/ee/src/workspace/workspace-factory.ts:190]: seems to work https://github.com/gitpod-io/gitpod/pull/13113#issuecomment-1255035538

**Comment by AlexTugarev** [comment]
Closing, as this was split in smaller PRs.

**Pull request closed without merging**